### PR TITLE
Use connection pool in HiveMetastoreClient

### PR DESCRIPTION
**Pull request by lgbo-ustc**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use connection pool for hive metastore client

1. remove lock for hive metastore client access
2. auto reconnect when connection is broken

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/Hive/HiveCommon.cpp [hunk 1]
```diff
@@ -192,53 +216,55 @@ HiveMetastoreClientFactory & HiveMetastoreClientFactory::instance()
...
+        };
+        auto client = std::make_shared<HiveMetastoreClient>(builder, context->getGlobalContext());
+        clients[name] = client;
+        return client;
     }
     return it->second;
 }
+const int HiveMetastoreClientFactory::conn_timeout_ms = 10000;
+const int HiveMetastoreClientFactory::recv_timeout_ms = 10000;
+const int HiveMetastoreClientFactory::send_timeout_ms = 10000;
```
Comment (kssenii) [inline src/Storages/Hive/HiveCommon.cpp:243]: Let's move to the top of the file.

File: src/Storages/Hive/HiveCommon.h [hunk 2]
```diff
@@ -128,13 +163,15 @@ class HiveMetastoreClientFactory final : private boost::noncopyable
...
+    static std::shared_ptr<Apache::Hadoop::Hive::ThriftHiveMetastoreClient> createThriftHiveMetastoreClient(const String & name);
+
 private:
     std::mutex mutex;
     std::map<String, HiveMetastoreClientPtr> clients;
 
-    const int conn_timeout_ms = 10000;
-    const int recv_timeout_ms = 10000;
-    const int send_timeout_ms = 10000;
+    const static int conn_timeout_ms;
```
Comment (kssenii) [inline src/Storages/Hive/HiveCommon.h:172]: why do you need it here? enough to have `static constexpr int conn_timeout_ms` in cpp file.

File: src/Storages/Hive/HiveCommon.h [hunk 3]
```diff
@@ -94,30 +125,34 @@ class HiveMetastoreClient : public WithContext
...
+
     LRUCache<String, HiveTableMetadata> table_metadata_cache;
-    mutable std::mutex mutex;
-    std::atomic<bool> expired{false};
+    ThriftHiveMetastoreClientPool client_pool;
 
     Poco::Logger * log = &Poco::Logger::get("HiveMetastoreClient");
+
+    const int max_retry = 3;
+    const UInt64 get_client_timeout = 1000000;
```
Comment (kssenii) [inline src/Storages/Hive/HiveCommon.h:155]: Same comment as above

**Review by kssenii — APPROVED** [review]

File: src/Storages/Hive/HiveCommon.cpp [hunk 4]
```diff
@@ -192,53 +226,52 @@ HiveMetastoreClientFactory & HiveMetastoreClientFactory::instance()
...
+    const auto & host = hive_metastore_url.getHost();
+    auto port = hive_metastore_url.getPort();
+
+    std::shared_ptr<TSocket> socket = std::make_shared<TSocket>(host, port);
+    socket->setKeepAlive(true);
+    socket->setConnTimeout(hive_metastore_client_conn_timeout_ms);
+    socket->setRecvTimeout(hive_metastore_client_recv_timeout_ms);
+    socket->setSendTimeout(hive_metastore_client_send_timeout_ms);
+    std::shared_ptr<TTransport> transport(new TBufferedTransport(socket));
+    std::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
```
Comment (kssenii) [inline src/Storages/Hive/HiveCommon.cpp:263]: why not use `std::make_shared`?

Comment (lgbo-ustc) [reply src/Storages/Hive/HiveCommon.cpp]: updated

**Pull request merged**